rasa/core/policies/flow_policy.py

```diff
@@ -470,53 +478,61 @@ def _select_next_action(
 
         tracker = tracker.copy()
 
         number_of_initial_events = len(tracker.events)
 
-        while not predicted_action or predicted_action.score == 0.0:
+        while isinstance(step_result, ContinueFlowWithNextStep):
             if not (current_flow := self.flow_stack.top_flow(self.all_flows)):
                 # If there is no current flow, we assume that all flows are done
                 # and there is nothing to do. The assumption here is that every
                 # flow ends with an action listen.
-                predicted_action = ActionPrediction(ACTION_LISTEN_NAME, 1.0)
-                break
-
-            if not (previous_step := self.flow_stack.top_flow_step(self.all_flows)):
-                raise FlowException(
-                    "The current flow is set, but there is no current step. "
-                    "This should not happen, if a flow is started it should be set "
-                    "to __start__ if it ended it should be popped from the stack."
+                step_result = PauseFlowReturnPrediction(
+                    ActionPrediction(ACTION_LISTEN_NAME, 1.0)
                 )
+            else:
+                with bound_contextvars(flow_id=current_flow.id):
+                    if not (
+                        previous_step := self.flow_stack.top_flow_step(self.all_flows)
+                    ):
+                        raise FlowException(
+                            "The current flow is set, but there is no current step. "
+                            "This should not happen, if a flow is started it should "
+                            "be set to __start__ if it ended it should be popped "
+                            "from the stack."
+                        )
 
-            structlogger.debug("flow.action.loop", previous_step=previous_step)
-            predicted_action = self._wrap_up_previous_step(
-                current_flow, previous_step, tracker
-            )
-            tracker.update_with_events(predicted_action.events or [], self.domain)
-
-            if predicted_action.action_name:
-                # if the previous step predicted an action, we'll stop here
-                # the step is not completed yet and we need to predict the
-                # action first before we can try again to wrap up this step and
-                # advance to the next one
-                break
+                    structlogger.debug(
+                        "flow.execution.loop", previous_step_id=previous_step.id
+                    )
+                    current_step = self._select_next_step(
+                        tracker, previous_step, current_flow
+                    )
 
-            current_step = self._select_next_step(tracker, previous_step, current_flow)
+                    if current_step:
+                        # this can't be an else, because the previous if might change
+                        # this to "not None"
 
```

Comment:
this comment is outdated, there are now other steps in between this if and the previous one and so an else wouldn't even be considered here